# Task 4.05

## Table of Contents
1. Import Libraries + Data
2. Data Consistency Checks
3. Missing Values
4. Duplicates
5. 4.05 Task
    - further investigation
    - mixed type
    - check for missing values
    - address missing values
    - check for duplicates
    - export

# 01. Import Libraries + Data

In [1]:
# Importing Libraries
import pandas as pd
import numpy as np
import os

In [2]:
# Importing Data
path = r'/Users/alliecarlisle/Desktop/CareerFoundry_Data_Analytics/Instacart Basket Analysis'

In [3]:
path

'/Users/alliecarlisle/Desktop/CareerFoundry_Data_Analytics/Instacart Basket Analysis'

In [5]:
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [4]:
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

# 02. Data Consistency Checks

In [6]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  time_between_orders  
count       3.421083e+06         3.214874e+06  
mean        1.345202e+01         1.111484e+01  
std         4.226088e+00         9.206737e+00  
min         0.000000e+00         0.000000e+00  
25%         1.000000e+01         4.000000e+00  
50%         1.300000e+01         7.000000e+00  
75%         1.600000e+01         1.500000e+01  
max         2.300000e+01         3.000000e+01

In [7]:
# Create a dataframe

df_test = pd.DataFrame()

In [8]:
# Create a mixed type column

df_test['mix'] = ['a', 'b', 1, True]

In [9]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [10]:
# checking for mixed types in df_test

for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [11]:
# fixing mixed data type in the mix column of df_test

df_test['mix'] = df_test['mix'].astype('str')

# 03. Missing Values

In [12]:
# find total # of missing values in df_prods

df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [13]:
# create a new dataframe with only missing values in product_name column

df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [14]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [15]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

In [16]:
df_prods.median()

/var/folders/w0/qzmf_lf542x0x6lpcdj7lnww0000gn/T/ipykernel_18716/3387230063.py:1: FutureWarning: The default value of numeric_only in DataFrame.median is deprecated. In a future version, it will default to False. In addition, specifying 'numeric_only=None' is deprecated. Select only valid columns or specify the value of numeric_only to silence this warning.
  df_prods.median()


product_id       24845.0
aisle_id            69.0
department_id       13.0
prices               7.1
dtype: float64

In [21]:
# handling missing values in product_name column of products.csv

df_prods.shape

(49693, 5)

In [19]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [20]:
df_prods_clean.shape

(49677, 5)

# 04. Duplicates

In [22]:
# looking for full duplicates in df_prods_clean

df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [23]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [24]:
# addressing duplicates

df_prods_clean.shape

(49677, 5)

In [25]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [26]:
df_prods_clean_no_dups.shape

(49672, 5)

# 05. 4.05 Task

## 2. Is there anything in df_ords that should be investigated further?

### Answer: It looks like there is at least one outlier in the time_between_orders column, as the max is 30 and the 75% IQR is 15. This could alter the mean to be higher than what makes sense.

In [31]:
df_ords.describe()

order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  time_between_orders  
count       3.421083e+06         3.214874e+06  
mean        1.345202e+01         1.111484e+01  
std         4.226088e+00         9.206737e+00  
min         0.000000e+00         0.000000e+00  
25%         1.000000e+01         4.000000e+00  
50%         1.300000e+01         7.000000e+00  
75%         1.600000e+01         1.500000e+01  
max         2.300000e+01         3.000000e+01

In [37]:
# drop unnamed column in orders_wrangled.csv
df_ords = df_ords.drop(columns = ['Unnamed: 0'])

KeyError: "['Unnamed: 0'] not found in axis"

In [33]:
df_ords.head()

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0   2539329        1             1          2                  8   
1   2398795        1             2          3                  7   
2    473747        1             3          3                 12   
3   2254736        1             4          4                  7   
4    431534        1             5          4                 15   

   time_between_orders  
0                  NaN  
1                 15.0  
2                 21.0  
3                 29.0  
4                 28.0

In [34]:
df_ords.describe()

order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  time_between_orders  
count       3.421083e+06         3.214874e+06  
mean        1.345202e+01         1.111484e+01  
std         4.226088e+00         9.206737e+00  
min         0.000000e+00         0.000000e+00  
25%         1.000000e+01         4.000000e+00  
50%         1.300000e+01         7.000000e+00  
75%         1.600000e+01         1.500000e+01  
max         2.300000e+01         3.000000e+01

## 3. Check for mixed-type data

In [35]:
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

## 5. Check for missing values

In [36]:
df_ords.isnull().sum()

order_id                    0
user_id                     0
order_number                0
order_dow                   0
order_hour_of_day           0
time_between_orders    206209
dtype: int64

### Answer - It looks like there are 206,209 missing values in the time_between_orders column. I'm assuming these signify orders where there was no prior order, so there is only one order for that customer.

## 6. Address the missing values

In [38]:
# create a new dataframe with only missing values in time_between_orders column

df_ords_nan = df_ords[df_ords['time_between_orders'].isnull() == True]

In [39]:
df_ords_nan.shape

(206209, 6)

In [40]:
df_ords.shape

(3421083, 6)

In [41]:
# create a new dataframe excluding the missing values in time_between_orders column
df_ords_clean = df_ords[df_ords['time_between_orders'].isnull() == False]

In [42]:
df_ords_clean.shape

(3214874, 6)

### Answer - As the missing values made up about 6% of the data, I decided to remove the missing data.

## 7. Check for duplicate values

In [44]:
# looking for full duplicates in df_ords_clean

df_ords_dups = df_ords_clean[df_ords_clean.duplicated()]

In [45]:
df_ords_dups

Empty DataFrame
Columns: [order_id, user_id, order_number, order_dow, order_hour_of_day, time_between_orders]
Index: []

### Answer - there are no duplicates

# 06. Export Data

In [46]:
# Export df_prods_clean_no_dups
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

In [47]:
# Export df_ords_clean
df_ords_clean.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))